proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2948,198 +3035,256 @@ lemma cancelIPC_ccorres1:
-      apply (rule ccorres_rhs_assoc)+
-      apply csymbr
-      apply csymbr
-      apply (rule ccorres_pre_getEndpoint)
-      apply (rule ccorres_assert)
-      apply (rule ccorres_symb_exec_r) \<comment> \<open>ptr_get lemmas don't work so well :(\<close>
+          apply (clarsimp simp: st_tcb_at'_def)
+          apply (frule (1) obj_at_cslift_tcb)
+          apply (clarsimp simp: typ_heap_simps ctcb_relation_BlockedOnSend_blockingObject)
+         apply ceqv
         apply (rule ccorres_symb_exec_r)
-          apply (simp only: fun_app_def simp_list_case_return return_bind ccorres_seq_skip)
+          apply (simp only: list_case_If)
           apply (rule ccorres_rhs_assoc2)
           apply (rule ccorres_rhs_assoc2)
           apply (rule ccorres_rhs_assoc2)
           apply (ctac (no_vcg) add: cancelIPC_ccorres_helper)
-            apply (ctac add: setThreadState_ccorres_valid_queues')
-           apply (wp hoare_vcg_all_lift set_ep_valid_objs' | simp add: valid_tcb_state'_def split del:if_split)+
-       apply (simp add: ThreadState_defs)
-         apply clarsimp
-         apply (rule conseqPre, vcg, rule subset_refl)
+            apply (rule_tac P="tcb_at' thread" in ccorres_cross_over_guard)
+            apply (rule_tac xf'=ret__unsigned_longlong_'
+                        and val="scast ThreadState_BlockedOnSend"
+                        and R="st_tcb_at' ((=) threadState) thread"
+                        and R'=UNIV
+                        in ccorres_symb_exec_r_known_rv)
+               apply (rule conseqPre, vcg)
+               apply (clarsimp simp: st_tcb_at'_def)
+               apply (frule (1) obj_at_cslift_tcb)
+               apply (fastforce simp: typ_heap_simps ctcb_relation_thread_state_to_tsType
+                                      thread_state_to_tsType_eq_BlockedOnSend)
+              apply ceqv
+             apply (simp only: ThreadState_defs)
+             apply ccorres_rewrite
+             apply (ctac add: setThreadState_ccorres)
+            apply vcg
+           apply wp
+           apply simp
+          apply (clarsimp simp: ThreadState_defs)
+          apply (frule (1) tcb_at_h_t_valid)
+          apply simp
+         apply vcg
         apply (rule conseqPre, vcg)
         apply clarsimp
-       apply clarsimp
-       apply (rule conseqPre, vcg, rule subset_refl)
-      apply (rule conseqPre, vcg)
+       apply vcg
+      \<comment> \<open>Restart\<close>
+      apply (simp add: word_sle_def ThreadState_defs ccorres_cond_iffs
+                 cong: call_ignore_cong,
+             rule ccorres_return_Skip)
+     \<comment> \<open>Post wp proofs\<close>
+     apply vcg
+    apply (rule_tac Q'="\<lambda>rv s. tcb_at' thread s
+                               \<and> st_tcb_at' ((=) threadState) thread s
+                               \<and> invs' s \<and> weak_sch_act_wf (ksSchedulerAction s) s
+                               \<and> sym_refs_asrt s"
+                in hoare_strengthen_post)
```

Comment:
hard to tell in this interface, but the `in` should right-align with the `_tac`

Reply:
Done, thanks.